### PR TITLE
v2

### PR DESCRIPTION
**Pull request by leon-do**
- not ERC20 standard to preserve privacy
- no events emitted
- balanceOf is only for msg.sender
- nonce is non-incremental
- solidity > v0.80

@HarryR

### CODE REVIEW

File: contracts/WROSE.sol [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      payable(msg.sender).transfer(reward);
+      payable(to).transfer(value);
+      return true;
+    }
+
+    function metaWithdrawHash(address to, uint256 value, uint256 nonce, uint256 reward) public view returns(bytes32){
+      return keccak256(abi.encodePacked(address(this), to, value, nonce, reward));
+    }
+
+    function getSigner(bytes32 _hash, bytes memory _signature) public pure returns (address) {
```
Comment (nhynes) [inline contracts/WROSE.sol]: Why not use https://docs.openzeppelin.com/contracts/4.x/api/utils#SignatureChecker? Even if it's the same, it cuts down on the new code to be audited.

File: contracts/WROSE.sol [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        balances[msg.sender] -= value;
+        payable(msg.sender).transfer(value);
+    }
+
+    function metaWithdraw(bytes memory signature, address to, uint256 value, uint256 nonce, uint256 reward) public returns (bool) {
+      bytes32 metaHash = metaWithdrawHash(to, value, nonce, reward);
+      address signer = getSigner(metaHash, signature);
+
+      require(balances[signer] >= value + reward, "Insufficient Balance");
+      require(signer != address(0), "Signer is 0x0");
```
Comment (nhynes) [inline contracts/WROSE.sol]:
```suggestion
      require(signer != address(0), "irrecoverable signature");
```
signer is 0 when the signature length is wrong or the recovery id is out of bounds. that's a lot to say in an error message, though.

File: contracts/WROSE.sol [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      require(replayNonce[nonce] == false, "Nonce has been used");
```
Comment (nhynes) [inline contracts/WROSE.sol]:
```suggestion
      require(!replayNonce[nonce], "Nonce has been used");
```

File: contracts/WROSE.sol [hunk 4]
```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.19;
+
+contract WROSE {
```
Comment (nhynes) [inline contracts/WROSE.sol:12]:
```suggestion
/// This WROSE allows meta-withdrawals to accounts that do not have gas.
/// Token transfers are not private because Sapphire compute node operators can see which accounts' storage slots are active.
/// Do NOT use this code for anything that requires privacy if you do not trust Sapphire compute node operators.
contract WROSE {
```

File: contracts/WROSE.sol (hunk 1 above)
Comment (leon-do) [reply contracts/WROSE.sol]: Great suggestion. implemented

File: contracts/WROSE.sol (hunk 2 above)
Comment (leon-do) [reply contracts/WROSE.sol]: Fixed.

File: contracts/WROSE.sol (hunk 4 above)
Comment (leon-do) [reply contracts/WROSE.sol:12]: Added comments

**Comment by leon-do** [comment]
any suggestions @harryR ?

File: contracts/WROSE.sol [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+contract WROSE {
+    using ECDSA for bytes32;
+
+    mapping(address => uint256) private balances;
+    mapping(uint256 => bool) public replayNonce;
+
+    function balanceOf() public view returns (uint256) {
+        return balances[msg.sender];
```
Comment (HarryR) [inline contracts/WROSE.sol:19]: Minor point, could be `balance` or `myBalance` to avoid confusion with `balanceOf(address)`.

File: contracts/WROSE.sol [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    function metaWithdraw(
+        bytes memory signature,
+        address to,
+        uint256 value,
+        uint256 nonce,
+        uint256 reward
+    ) public returns (bool) {
+        bytes32 metaHash = metaWithdrawHash(to, value, nonce, reward);
+        address signer = metaHash.toEthSignedMessageHash().recover(signature);
```
Comment (HarryR) [inline contracts/WROSE.sol]: I have a strong preference for using typed structure signing as specified in the EIP-712 standard which uses with `eth_signTypedData`.

My concern is that if wrose.io was hijacked by a malicious entity they could change the app to change the `value`, `reward` and `to` parameters behind the scenes, and the user will not be able to identify this change because all they see in the UI is random looking hex encoded string.

File: README.md [hunk 7]
```diff
@@ -29,7 +29,7 @@ truffle migrate --network sapphire-mainnet
 ```
 =====================
 
-   Deploying 'WROSE9'
+   Deploying 'WROSE'
```
Comment (HarryR) [inline README.md:28]: The contract hasn't been re-deployed though...

File: contracts/WROSE.sol [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    using ECDSA for bytes32;
+
+    mapping(address => uint256) private balances;
+    mapping(uint256 => bool) public replayNonce;
+
+    function balanceOf() public view returns (uint256) {
+        return balances[msg.sender];
+    }
+
+    function deposit() public payable {
```
Comment (HarryR) [inline contracts/WROSE.sol:22]: Any chance of a `depositTo` function to allow people to deposit into another persons account?

**Review by HarryR — COMMENTED** [review]
Hi, here is some feedback

File: contracts/WROSE.sol (hunk 6 above)
Comment (leon-do) [reply contracts/WROSE.sol]: You're right. Let me made adjustments for EIP-712.

Very valid points.

File: contracts/WROSE.sol (hunk 8 above)
Comment (leon-do) [reply contracts/WROSE.sol:22]: The idea is 
`deposit`   moves ROSE from wallet to contract
`withdraw` moves ROSE from contract to original wallet
`metaWithdraw` moves ROSE from contract to any wallet.

TLDR: metaWithdraw is the same as `depositTo`

File: README.md (hunk 7 above)
Comment (leon-do) [reply README.md:28]: This is just an example

File: contracts/WROSE.sol (hunk 6 above)
Comment (leon-do) [reply contracts/WROSE.sol]: @HarryR 

Updated to use EIP-712

Screenshot 2023-05-17 at 4 41 55 PM

**Comment by nhynes** [comment]
When this gets merged, I'm going to use it as the basis for my in-game currency.

**Comment by HarryR** [comment]
> When this gets merged, I'm going to use it as the basis for my in-game currency.

I am not confident in its privacy properties yet, in its current state it only offers superficial privacy.

 1. Relayer can observe transaction content.
 2. Storage I/O links transactions together via access to `balances`.
 3. It doesn't support internal transfers.
 4. It's also griefable, in the sense that I can spam the relayer with fake transactions to waste gas

Good things though:

 * No admin backdoor
 * No admin enforced relayer (so new relayer can be made if official relayer goes down)

File: contracts/WROSE.sol [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+    function getSigner(bytes memory _signature, address _to,uint256 _value,uint256 _nonce,uint256 _reward)
+        private
+        view
+        returns (address)
+    {
+        // EIP721 domain type
+        string memory name = "WROSE";
+        string memory version = "1";
+        uint256 chainId = 23294;
```
Comment (HarryR) [inline contracts/WROSE.sol:61]: chainId shouldn't be hard coded.

**Comment by leon-do** [comment]
> * Relayer can observe transaction content.

Not sure how to solve this. Need to brew on it.

> * Storage I/O links transactions together via access to `balances`.

Can you please provide more info on this? Also leaning on @nhynes for his thoughts.

> * It doesn't support internal transfers.

Correct. But will compromise privacy?

> * It's also griefable, in the sense that I can spam the relayer with fake transactions to waste gas

There are currently some basic checks but this can be solved with robust relay logic. If the transaction is bad, don't broadcast.

**Comment by HarryR** [comment]
To avoid Storage I/O link-ability the general idea is converting some (or all) of an accounts balance into a 'Coupon'; think of it like a bearer bond - the holder knows it's valid because it's signed by the contract & the contract knows it's unspent because the `nonce` isn't present in its storage, upon presentation of the coupon the contract will credit the holder with its balance. If the coupon is lost the value it holds will forever be in limbo.

The `nonce` storage key isn't accessed until somebody tries to spend it, the only information that can be observed from watching storage I/O patterns is if somebody tries to spend the same coupon twice.

e.g.

 * the contract generates a long-term keypair in the constructor, then publishes its public key
 * sender invokes 'convert to coupon', contract subtracts coupon amount from balance and signs it
 * recipient invokes 'redeem coupon', contract validates signature then adds hash of signature to the 'prevent double-spend set'

This raises the question of disputes, if I give you a coupon and you can't spend it - you say I must have spent it, or from my perspective are you lying and saying that you couldn't spend the coupon when you really did? So we need to make the coupon only spendable by the recipient's account (then any dispute becomes - you gave me the wrong account, I have proof of an unspent coupon for that account)

Sapphire has builtins for authenticated encryption:

```solidity
    address private constant DECRYPT = 0x0100000000000000000000000000000000000004;

    function _decrypt(bytes32 key, bytes32 nonce, bytes calldata ciphertext, bytes memory additionalData)
        private view
        returns (bytes memory)
    {
        (bool success, bytes memory plaintext) = DECRYPT.staticcall(
            abi.encode(key, nonce, ciphertext, additionalData)
        );

        require(success);

        return plaintext;
    }
```

**Comment by nhynes** [comment]
I think that issuing cashier's checks is a viable strategy. You might want to either only issue fixed denominations, or else reconstruct a utxo system to make redemptions more private. It's a good UX questions whether people will tolerate this approach, but it can't be more complicated than tcash.

Is this a good summary of the approach?
1. `requestChecks(denominations) external` - creates some checks that will be given to the requester in the next block
    * checks are delayed by one block to prevent a requester who is also a compute node operator from observing the response value as a side effect (e.g., encrypted event body)
     * this operation will transfer the total amount being converted 
2. `collectChecks() external returns (Check[] memory checks)` - returns the checks that have been issued to the requester
3. `cashCheck(Check calldata check, address depositAccount) external` - marks the check as cashed, crediting the deposit account from the bank balance

**Comment by leon-do** [comment]
These are great suggestions. It should probably be it's own branch.

@nhynes would your users be ok with this UX?

**Comment by nhynes** [comment]
My users are okay with both approaches :) I would prefer the tickets one, as it saves me writing a separate tickets contract :)
